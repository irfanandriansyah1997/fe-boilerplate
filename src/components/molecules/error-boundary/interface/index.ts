import {
  ErrorBoundaryPropsWithRender,
  FallbackProps
} from 'react-error-boundary';

/**
 * Error Boundary Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.11.19
 */
export type IErrorBoundaryProps = Omit<
  ErrorBoundaryPropsWithRender,
  'fallbackRender'
>;

/**
 * Error Fallback Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.11.19
 */
export interface IErrorFallbackProps extends FallbackProps {
  canReset?: boolean;
}
