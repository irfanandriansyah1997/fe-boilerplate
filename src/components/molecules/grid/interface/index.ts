import { FC } from 'react';

/**
 * Grid Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.30
 */
export interface IGridProps {
  Cell: FC<ICellProps>;
  column?: number;
  row?: number;
}

/**
 * Cell Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.30
 */
export interface ICellProps {
  column: number;
  row: number;
}
