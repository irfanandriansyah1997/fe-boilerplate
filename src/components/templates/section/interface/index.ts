export * from './section-context.interface';
export * from './section-hooks.interface';
