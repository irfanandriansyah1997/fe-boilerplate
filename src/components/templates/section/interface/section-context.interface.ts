/**
 * Section Context Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.27
 */
export interface ISectionContextState {
  markdownURL?: string;
}

/**
 * Section Context Action Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.27
 */
export interface ISectionContextAction {
  setMarkdownURL(string: string | undefined): void;
}

/**
 * Section Context Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.27
 */
export interface ISectionContext {
  action: ISectionContextAction;
  state: ISectionContextState;
}
