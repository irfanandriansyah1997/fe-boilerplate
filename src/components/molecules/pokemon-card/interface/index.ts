import { FC } from 'react';

import { IPokemonDetail } from '../../../../interface/pokemon';

/**
 * Pokemon Card Fallback Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.11.19
 */
export interface IPokemonCardFallbackProps {
  name: string;
}

/**
 * Pokemon Card Type Enum
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.11.21
 */
export enum IPokemonCardTypeEnum {
  basic = 0,
  lazyLoadImage = 1
}

/**
 * Pokemon Card Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.11.19
 */
export interface IPokemonCardProps extends IPokemonDetail {
  type?: IPokemonCardTypeEnum;
}

/**
 * Pokemon Card Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.11.19
 */
export type IPokemonCard = FC<IPokemonCardProps> & {
  Fallback: FC<IPokemonCardFallbackProps>;
};
