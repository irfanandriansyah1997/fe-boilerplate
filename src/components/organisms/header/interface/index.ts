import { IMenuItem } from '../../../../interface/component';

/**
 * Header Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.14
 */
export interface IHeaderProps {
  menu: IMenuItem[];
}
