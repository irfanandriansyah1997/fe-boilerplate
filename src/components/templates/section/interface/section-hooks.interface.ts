import { NullAble } from '../../../../interface/general';

/**
 * Section Hooks Dispatch Parameter Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.27
 */
export type ISectionHooksDispatchParameter =
  | NullAble<string>
  | ((param: NullAble<string>) => NullAble<string>);

/**
 * Section Hooks Dispatch Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.27
 */
export type ISectionHooksDispatch = (
  param: ISectionHooksDispatchParameter
) => void;

/**
 * Section Hooks Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.27
 */
export type ISectionHooks = [NullAble<string>, ISectionHooksDispatch];
