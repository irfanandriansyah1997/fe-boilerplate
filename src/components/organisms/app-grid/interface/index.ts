import { FC } from 'react';

import { ICellProps, IGridProps } from '../../../molecules/grid/interface';

/**
 * App Grid Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.30
 */
export interface IAppGridProps extends IGridProps {
  Cell: FC<ICellProps>;
  onForceRerender(): void;
  onUpdateGrid(): void;
}
