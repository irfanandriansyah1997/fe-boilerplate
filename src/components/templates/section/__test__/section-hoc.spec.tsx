import { act, fireEvent, render, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import '../../../organisms/markdown-reader/__test__/markdown-reader-mock.spec';
import { simulateResize } from '../../../../helper';
import { IMenu } from '../../../../interface/component';
import SectionTemplates from '..';
import { MOCK_FETCH_SECTION, MockComponentSection } from './section-mock.spec';

const MOCK_MENU: IMenu = {
  menu: [
    {
      text: `Wrap Use Memo`,
      to: `/context/part-1`
    },
    {
      text: `Separate Context Value`,
      to: `/context/part-2`
    },
    {
      text: `Colocate State`,
      to: `/context/part-3`
    }
  ],
  subtitle: `Sample Subtitle`,
  title: `Sample Title`
};

const pushSpy = jest.fn();

jest.mock(`react-router-dom`, () => ({
  ...jest.requireActual(`react-router-dom`),
  useHistory: () => ({
    push: pushSpy
  })
}));

describe(`Testing Section HOC Render On Browser`, () => {
  const tempFetch = global.fetch;

  beforeEach(() => {
    global.fetch = jest.fn().mockImplementation(MOCK_FETCH_SECTION);
  });

  it(`Snapshot Testing`, async () => {
    const Component = SectionTemplates(MockComponentSection, MOCK_MENU);

    const { container, getByTestId, getByText } = render(
      <MemoryRouter initialEntries={[`/context/part-1`]}>
        <Component />
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(getByTestId(/markdown-code/)).toBeInTheDocument();
      expect(fetch).toHaveBeenCalledWith(`sample-url`);
      expect(getByText(/^Hello World$/i)).toBeInTheDocument();
    });

    expect(container).toMatchSnapshot();
  });

  it(`Simulate User Click Back To Homepage`, async () => {
    const Component = SectionTemplates(MockComponentSection, MOCK_MENU);

    const { getByTestId, getByText } = render(
      <MemoryRouter initialEntries={[`/context/part-1`]}>
        <Component />
      </MemoryRouter>
    );

    /**
     * Simulate Resize Desktop Size
     */
    await act(simulateResize(1366));

    await waitFor(() => {
      expect(getByTestId(/markdown-code/)).toBeInTheDocument();

      /**
       * Expect Data From Local Storage Not Fetch API
       */
      expect(localStorage.getItem).toHaveBeenCalledWith(`sample-url`);
      expect(fetch).not.toHaveBeenCalled();
    });

    /**
     * Simulate User Click Back Button
     */
    const buttonBack = getByText(/arrow_back/);
    fireEvent.click(buttonBack);

    expect(pushSpy).toHaveBeenCalledWith(`/`);
  });

  afterEach(() => {
    jest.clearAllMocks();

    global.fetch = tempFetch;
  });
});
