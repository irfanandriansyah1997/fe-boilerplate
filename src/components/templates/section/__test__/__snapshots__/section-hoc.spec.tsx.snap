// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Section HOC Render On Browser Snapshot Testing 1`] = `
<div>
  <div>
    <div
      data-testid="sidebar-msite"
    >
      <div
        class="o-sidebar-mobile__navbar"
        data-testid="sidebar-msite-navbar"
      >
        <button
          class="material-icons"
          data-testid="sidebar-msite-navbar-button"
          type="button"
        >
          menu
        </button>
        <h6
          data-testid="sidebar-msite-navbar-title"
        >
          Sample Title
        </h6>
      </div>
    </div>
    <div
      class="t-section-item t-section-item--with-markdown"
    >
      <div
        class="t-section-item__content"
      >
        <div>
          Hello World
        </div>
      </div>
      <div
        class="t-section-item__markdown"
      >
        <div
          data-testid="markdown-code"
        >
          <div>
            ### Hello World 
 Sample Text On Markdown
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
